File: Clases/Lab08/UniquePath.py
## UniquuOAth.py

def uniquePathsWithObstacles(self, obstacleGrid):
    """
    :type obstacleGrid: List[List[int]]
    :rtype: int
    """
    ## Creamos la estrcutura que nos ayudara
    caminos = [ [ 0 for y in range( len(obstacleGrid) ) ] for x in range( len(obstacleGrid[0]) ) ]
    
    for j in range(len(obstacleGrid)):
        for i in range(len(obstacleGrid[j])):
            
                if (i-1 < 0 and j-1 < 0):
                    caminos[i][j] = 1
                    continue
                ## lateral izquierdo
                elif (i-1 < 0):
                    if(obstacleGrid[i][j-1] != 1):
                        # caminos[i][j] = caminos[i-1][j]
                        caminos[i][j] = caminos[i][j-1]
                    else:
                        caminos[i][j] = 0
                ## Superior
                elif(j-1 < 0):
                    if(obstacleGrid[i-1][j] != 1):
                        caminos[i][j] = caminos[i-1][j]
                    else:
                        caminos[i][j] = 0
                else:
                    ## casos posibles 
                    if(obstacleGrid[i-1][j] == 1 ):
                        caminos[i][j] = caminos[i][j-1]
                    elif(obstacleGrid[i][j-1] == 1):
                        caminos[i][j] = caminos[i-1][j]
                    elif(obstacleGrid[i-1][j] == 1 and obstacleGrid[i][j-1] == 1):
                        caminos[i][j] = 0
                    else:
                        caminos[i][j] = caminos[i-1][j] + caminos[i][j-1]
    return caminos
                        
a = uniquePathsWithObstacles(None, [[0,0,0],[0,1,0],[0,0,0]])
print(a)

File: Clases/Lab08/UniquePath.java
package Clases.Lab08;

// UniquePath.java
// @author: Paul Luque
// @description: Encontrar el numero de caminos para llegar el otro extremo de la matris, en la matris de obtavulos, el 1 es un obtaculo y no podemos pasar por ese lugar

public class UniquePath {
    public static void main(String[] args) {
        int[][] test = { { 0, 0, 0, 1 }, { 0, 1, 0, 1 }, { 0, 0, 0, 0 } };
        int resultado = 0;
        resultado = uniquePathsWithObstacles(test);
        System.out.println(resultado);
    }

    public static int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int[][] caminos = new int[obstacleGrid.length][obstacleGrid[0].length];
        for (int j = 0; j < obstacleGrid.length; j++) {
            for (int i = 0; i < obstacleGrid[j].length; i++) {
                // System.out.println(obstacleGrid[j][i]);

                // Verificacmos si esta bloqueado
                if (obstacleGrid[j][i] == 1) {
                    caminos[j][i] = 0;
                    continue;// finalizar
                }

                // Inicio del cuadro
                if (i - 1 < 0 && j - 1 < 0) {
                    caminos[j][i] = 1;
                    // lateral izquierdo
                } else if (i - 1 < 0) {
                    caminos[j][i] = caminos[j - 1][i]; // Sumamos el de arriba
                    // Superior
                } else if (j - 1 < 0) {
                    caminos[j][i] = caminos[j][i - 1]; // Sumamos el de la izquierda
                    // Cuadro normal
                } else {
                    caminos[j][i] = caminos[j - 1][i] + caminos[j][i - 1];

                }

            }

        }

        // imprimimos la matris "caminos"
        for (int j = 0; j < caminos.length; j++) {
            for (int i = 0; i < caminos[j].length; i++) {
                System.out.print(caminos[j][i] + " ");
            }
            System.out.println();
        }

        return caminos[obstacleGrid.length - 1][obstacleGrid[0].length - 1];
    }

}